### PR TITLE
Deployment info obtained from the url

### PR DESCRIPTION
**Pull request by pramodcs**

### CODE REVIEW

**Comment by pramodcs** [comment]
I won't be surprised if some of the abstractions needs to be refined. Let me know what you think. It seems to be working fine from an end user's standpoint.

File: ushahidi/src/main/java/com/ushahidi/android/data/api/service/SiteConfigAPI.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Intended to be used to get the title of the url (and also a way to verify that the url is a valid
+ * ushahidi platform url) at the time the user enters/updates the deployment information.
+ *
+ * @author Ushahidi Team <team@ushahidi.com>
+ */
+public interface SiteConfigAPI {
+
+    @GET("/api/v3/config/site")
+    void getConfig(Callback<Config> response);
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/data/api/service/SiteConfigAPI.java]: I will return an `Observable` here. Then I can make a composition that actually fetches the data via the API then saves the details to the db.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+import com.google.gson.annotations.SerializedName;
+
+import java.util.List;
+
+/**
+ * Deployment Site Configuration Entity
+ *
+ * @author Ushahidi Team <team@ushahidi.com>
+ */
+public class Config {
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: Instead of creating a new model for this, you should add the extra properties to the existing deployment [entity](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/domain/entity/Deployment.java) as it's the same thing.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    
+    @SerializedName("id")
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: Plus the retrofit specific annotations should be defined for the `data` entities. Look at the data entities for say [PostEntity](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/data/entity/PostEntity.java)

File: ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/UpdateDeploymentFragment.java [hunk 4]
```diff
@@ -67,6 +73,24 @@
 
     private DeploymentModel mDeploymentModel;
 
+    private Callback<Config> mSiteConfigCallback = new Callback<Config>() {
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/UpdateDeploymentFragment.java]: We shouldn't be doing this in the view. The presenter should handle this.

**Comment by eyedol** [comment]
This is great for a first start hooking into the overall architecture and seems to be working however it deviates a lot from the core architecture which makes it hard to follow and also hard to write test for.

Amend the usecase for [`GetDeploymentUsecase`](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/domain/usecase/deployment/GetDeploymentUsecase.java). Add a property that asks where to fetch the data from. `Online` or `Database` Follow how `GeoJson` data is fetched.

We can pair program on this to get you a firm understanding of things. Let me know.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java (hunk 2 above)
Comment (pramodcs) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: Ok. I thought deployment entity was meant for a different endpoint. Will make the changes.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java (hunk 3 above)
Comment (pramodcs) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: I am not sure what you mean by data entities.

**Comment by pramodcs** [comment]
I shall make some changes and then we can do pair program.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java (hunk 3 above)
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: They're are the [entities](https://github.com/ushahidi/platform-android/tree/master/ushahidi/src/main/java/com/ushahidi/android/data/entity) defined in the [`data`](https://github.com/ushahidi/platform-android/tree/master/ushahidi/src/main/java/com/ushahidi/android/data) package.

**Comment by eyedol** [comment]
@pramodcs any update on this?. Let me know if you want us to work on this together.

File: ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java (hunk 3 above)
Comment (pramodcs) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: The Deployment entity is in the domain package itself. I am not sure what I should do for retrofit specific annotations.

Comment (pramodcs) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: I suppose you mean I should change both DeploymentEntity (in the data package) and Deployment in the domain package. I think you also mean that for parsing the response I should be using the one in the data package.

Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/entity/Config.java]: > The Deployment entity is in the domain package itself. I am not sure what I should do for retrofit specific annotations.

Move the retrofit annotation specifics to the DeploymentEntity in the data package. They shouldn't be in the domain one.

> I suppose you mean I should change both DeploymentEntity (in the data package) and Deployment in the domain package. I think you also mean that for parsing the response I should be using the one in the data package.

Correct.

**Comment by pramodcs** [comment]
Alright! Take a look once again @eyedol

File: ushahidi/src/main/java/com/ushahidi/android/data/entity/mapper/DeploymentEntityDataMapper.java [hunk 5]
```diff
@@ -75,7 +75,7 @@ public DeploymentEntity map(Deployment deployment) {
         if (deployment != null) {
             deploymentEntity = new DeploymentEntity();
             deploymentEntity._id = deployment._id;
-            deploymentEntity.setTitle(deployment.getTitle());
+            deploymentEntity.setName(deployment.getTitle());
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/data/entity/mapper/DeploymentEntityDataMapper.java]: For consistency sake, could we maintain the deployment title? Because it's heavily used.

File: ushahidi/src/main/java/com/ushahidi/android/data/repository/datasource/deployment/DeploymentApiDataSource.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    public DeploymentApiDataSource(@NonNull DeploymentApi deploymentApi,
+                                DeploymentDatabaseHelper deploymentDatabaseHelper) {
+        mDeploymentApi = deploymentApi;
+        mDeploymentDatabaseHelper = deploymentDatabaseHelper;
+    }
+
+    @Override
+    public Observable<List<DeploymentEntity>> getDeploymentEntityList() {
+        throw new UnsupportedOperationException();
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/data/repository/datasource/deployment/DeploymentApiDataSource.java]: Great. Good to see you added the exception for all unsupported operations.

File: ushahidi/src/main/java/com/ushahidi/android/domain/usecase/deployment/GetDeploymentUsecase.java [hunk 7]
```diff
@@ -18,27 +18,53 @@
...
      * Default constructor
      *
-     * @param deploymentId         The deployment Id
      * @param deploymentRepository The deployment repository
      * @param threadExecutor       The thread executor
      * @param postExecutionThread  The post execution thread
      */
     @Inject
-    protected GetDeploymentUsecase(Long deploymentId, DeploymentRepository deploymentRepository,
+    protected GetDeploymentUsecase(DeploymentRepository deploymentRepository,
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/usecase/deployment/GetDeploymentUsecase.java]: Ideally use-cases should do one thing. Right now it's fetching by URL as well. It make sense to give it own use case. Similar to this current implementation but only setDeploymentURL `FetchDeploymentUsecase`. Then set the URL minus the ID.

File: ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java [hunk 8]
```diff
@@ -46,12 +47,11 @@
  */
 public class AddDeploymentFragment extends BaseFragment implements AddDeploymentView {
 
-    @Bind(R.id.add_deployment_title)
-    EditText title;
-
     @Bind(R.id.add_deployment_url)
     EditText url;
 
+    ProgressDialog mProgressDialog;
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java]: Why don't we set the progress on the button like we do for the login?. It's a much better experience and it will be consistent with the rest of the app.

**Comment by eyedol** [comment]
@pramodcs Thanks for the update. This is looking really good. I added few inline comments to polish it up. Fetch Deployment URL I think should have it own use-case. Then with an accompanying test.

File: ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java (hunk 8 above)
Comment (pramodcs) [inline ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java]: My opinion is that _ProgressDialog_ provides a better UX by creating a translucent display around the dialog box making it clear for the user that he should not be engaging with it.

**Comment by pramodcs** [comment]
I have applied the changes you proposed except for the _ProgressBar_. If you think it is important let me know and I shall do the change.

File: ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java (hunk 8 above)
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/presentation/view/ui/fragment/AddDeploymentFragment.java]: I totally understand your point but I think displaying the progress on the button communicates the same message to the user but in a less annoying way. For consistency sake, could we keep the progress display on the button?

File: ushahidi/src/main/java/com/ushahidi/android/data/api/DeploymentApi.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package com.ushahidi.android.data.api;
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/data/api/DeploymentApi.java]: Missing license [header](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/presentation/presenter/deployment/ActivateDeploymentPresenter.java#L1-L15)

File: ushahidi/src/main/java/com/ushahidi/android/data/repository/datasource/deployment/DeploymentApiDataSource.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
+package com.ushahidi.android.data.repository.datasource.deployment;
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/data/repository/datasource/deployment/DeploymentApiDataSource.java]: Missing license [header doc](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/presentation/presenter/deployment/ActivateDeploymentPresenter.java#L1-L15)

File: ushahidi/src/main/java/com/ushahidi/android/domain/usecase/deployment/FetchDeploymentUsecase.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
+package com.ushahidi.android.domain.usecase.deployment;
```
Comment (eyedol) [inline ushahidi/src/main/java/com/ushahidi/android/domain/usecase/deployment/FetchDeploymentUsecase.java]: Missing license [header doc](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/presentation/presenter/deployment/ActivateDeploymentPresenter.java#L1-L15)

**Comment by eyedol** [comment]
@pramodcs Looking great. Thank you.

**Comment by eyedol** [comment]
The app crashes when you enter an invalid Ushahidi API URL. Say http://www.ushahidi.com. The reason is, the [ErrorMessageFactory](https://github.com/ushahidi/platform-android/blob/master/ushahidi/src/main/java/com/ushahidi/android/presentation/exception/ErrorMessageFactory.java#L102) class try to catch only IOException for errors thrown by retrofit. For some reason, retrofit error try to de-serialize the plain error response as a JSON string. Could you catch `Exception` instead and then return the default error message? I'll add a follow up issue so we can return an appropriate error message for this kind of case.

Also, not sure if this will be useful. When you attempt to add a deployment, the title field is in focus and ready for an input. Since we're asking for a valid URL, wouldn't it be nice if we already initialize the field with `http://`? Similar to how we do it when you touch the input field to type.

Once you handle the error message, it should be good for a merge. If you could look into [squashing](http://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit) your branch commits into a single commit, that would be nice. Eg. `git rebase -i master` then squash all the other commits except maybe the first one.
